### PR TITLE
fix: email-address with apostrophe not accepted

### LINKED ISSUE #15471: [CAL-4159] Email address with ' (apostrophe) not accepted
### Issue Summary

I've had someone try and use my scheduling page to book in but they couldn't because an email address with an ' (apostrophe) is not accepted as a valid email address. 

### Steps to Reproduce

1. Go to a scheduling page
2. Pick a date and time
3. Add an email address i.e. john.o'shean@domain.com
4. Address is not accepted

### Actual Results

- Email address not accepted

### Expected Results

- It should accept this email address as it is valid according to RFC5321/5322

### Technical details
Here's a regex that will correctly match all email addresses (according to ChatGPT)
```
^[a-zA-Z0-9.!#$%&'*+/=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$
```

### Evidence

image

<sub>[CAL-4159](https://linear.app/calcom/issue/CAL-4159/email-address-with-apostrophe-not-accepted)</sub>

### PR DESCRIPTION
**Pull request by RajdeepDs**
## What does this PR do?

- Fixes #15471 

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

### CODE REVIEW

**Comment by RajdeepDs** [comment]
Hi @keithwillcode,
I'm noticing that the tests for this pull request are failing. I'm not entirely familiar with the specific tests required. Could you provide some guidance on how to address these failures?

**Comment by hariombalhara** [comment]
This one seems to be fixed in later zod versions. I think upgrading zod should fix it everywhere? It seems we haven't updated in a while. So, it is due anyway.

https://github.com/colinhacks/zod/pull/2889

**Review by hariombalhara — CHANGES_REQUESTED** [review]
@RajdeepDs Thanks for the PR 🙏

Zod latest version should have this fixed already.  

But I am also thinking that it might make sense to have this flexibility of having email validation changed by changing the RegExp. It can allow us to fix email validation everywhere without heavily relying on zod's updates.


cc @keithwillcode @zomars what do you guys think ?

File: apps/api/v1/lib/validations/attendee.ts [hunk 1]
```diff
@@ -14,7 +15,7 @@ export const schemaAttendeeBaseBodyParams = Attendee.pick({
 const schemaAttendeeCreateParams = z
   .object({
     bookingId: z.number().int(),
-    email: z.string().email(),
+    email: z.string().regex(emailRegex),
```
Comment (hariombalhara) [inline apps/api/v1/lib/validations/attendee.ts:18]: We should be more declarative here like defining an `emailSchema` and importing that directly.

Comment (RajdeepDs) [reply apps/api/v1/lib/validations/attendee.ts]: Sure, I've made changes to the code by creating a new file `emailSchema.ts` that defines the `emailSchema` using the Reg Exp. This schema is now imported and used in the places where email validation is needed.

**Comment by zomars** [comment]
Agreed with @hariombalhara we need to make sure that updateProfile and signup are all using the same regex

**Comment by RajdeepDs** [comment]
@zomars
I have already ensured that updateProfile and signup are all using the same regex.

File: packages/lib/emailSchema.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+import { z } from "zod";
+
+export const emailRegex =
```
Comment (hariombalhara) [inline packages/lib/emailSchema.ts:3]: What is the source of this regex? Let's add it in the code.

Also, if that source doesn't have tests, we need the unit tests in here.

**Review by hariombalhara — DISMISSED** [review]
There are some occurences in app-store that aren't using it.

image


Also, I just noticed that the generated schema from zod is using zod.email(), we need to change that triple slash comment as well.
Screenshot 2024-07-23 at 11 48 54 AM

Screenshot 2024-07-23 at 11 49 20 AM

File: packages/lib/emailSchema.ts (hunk 2 above)
Comment (RajdeepDs) [reply packages/lib/emailSchema.ts:3]: The regex is sourced from https://github.com/colinhacks/zod/issues/3155#issuecomment-2060045794

**Comment by RajdeepDs** [comment]
@hariombalhara 
I have made the changes as requested. However, I couldn't figure out what we should use instead of `/// @zod.email()`. Before committing the changes, I would appreciate your suggestion on this.

**Comment by Amit91848** [comment]
@RajdeepDs You can import the email schema in `/packages/prisma/zod-utils.ts`

And then instead of `///@zod.email()` it would be `///@zod.custom(imports.emailSchema)`

If there are still some doubts you can checkout how zod.custom is being used in other places in `schema.prisma`

 https://github.com/calcom/cal.com/blob/a997edd8dbf77cd4ec5b61f9fe1ce34aaa291c2a/packages/prisma/zod-utils.ts#L112
 
 https://github.com/calcom/cal.com/blob/a997edd8dbf77cd4ec5b61f9fe1ce34aaa291c2a/packages/prisma/schema.prisma#L92-L93

**Review by zomars — DISMISSED** [review]
thanks again for the patience. Everything looks good from here. 🙏

**Review by zomars — DISMISSED** [review]

**Review by zomars — APPROVED** [review]

**Pull request merged**